### DISCUSSION TITLE
MET Rejecting All Obs in Prepbufr FIle

### DISCUSSION THREAD

**Discussion by craigr3** [dtcenter/METplus#2036] [Statistical Computation] [answered]
We have sent to you two prepbufr files.  One of them will use the observations and the other MET rejects all the observations.  Visual inspection of the files doesn't show an obvious difference.  We need to use the version that starts with prepbufr so need to find out why MET rejects the observations in that version.

**Comment by JohnHalleyGotway** [comment]
Hi Bob, I don't see those files that you're referencing. I checked both the RAL ftp site and my email for a DropBox notification but don't see either. Can you please point me to those files and, if possible, also share you PB2NC configuration file?

**Comment by craigr3** [comment]
They were sent to ***@***.*** using dodsafe.  What is your email? I can resend there.

**Reply by JohnHalleyGotway** [reply]
Bob, the last DoD SAFE email I see was from 12/9/22. And I assume this would be more recent than that.

It looks like the email address you listed above was sanitized somehow. You can send them to me at johnhg AT ucar.edu (johnhg@ucar.edu). I spelled it out so that it won't automatically get sanitized.

**Comment by craigr3** [comment]
We resent the files to johnhg@ucar.edu

**Reply by JohnHalleyGotway** [reply]
Thanks Bob. I grabbed the files and see that they're both a reasonable size:
```
-rw-r--r-- 1 johnhg rap 26M Feb  8 09:57 2023020400.nc
-rw-r--r-- 1 johnhg rap 24M Feb  8 09:57 prepbufr.2023020300.nc
```
Then I ran a python script to dump them out to ascii to make it easier to look for diffs:
```
python3 MET-develop/scripts/utility/print_pointnc2ascii.py prepbufr.2023020300.nc > prepbufr.2023020300.txt
python3 MET-develop/scripts/utility/print_pointnc2ascii.py 2023020400.nc > 2023020400.txt
```
Inspecting those ascii files, I don't see huge differences. But here are some potential ones:

1. The most obvious diff is in naming. One begins with **prepbufr** while the other does not.
   - Confirm that your runtime script is actually finding the expected filename and passing it to Point-Stat.
2. The list of message types differ slightly:
   - prepbufr.2023020300.txt has ADPSFC, ADPUPA, SFCSHP, VADWND.
   - 2023020400.txt has ADPSFC, ADPUPA, SFCSHP.
   - Check if the differences are due to the absence of **VADWND** obs in the 2/4/2023 data.
3. I did find some embedded whitespace in some station names that appears in both files: **Lod Airp**
   - I'm not sure if the embedded whitespace is causing a problem but it does appear in both files. So presumably it wouldn't cause a *difference* in behavior.
4. The list of variable names differ slightly:
   - prepbufr.2023020300.txt has DPT, HGT, MIXR, PRES, RH, SPFH, TMP.
   - 2023020400.txt has DPT, HGT, PRES, RH, SPFH, TMP.
   - Check if the differences are due to the absence of **MIXR** obs in the 2/4/2023 data.
5. Visual inspection of the lat/lon values reveals no red flags. But I ran plot_point_obs to get a better picture of them. I don't see any obvious problems in location.
<img width="1140" alt="Screen Shot 2023-02-08 at 10 59 34 AM" src="https://user-images.githubusercontent.com/21087144/217613620-8e5a7389-c285-4f78-b7fb-107fb7288344.png">

In summary, I see no obvious red flags but hopefully the smaller diffs give you some things to explore.

**Reply by craigr3** [reply]
Did you run them through point_stat?  The one with out the prepbufr in the name runs fine through point_stat with not all the obs rejected.  The one that starts with prepbufr when run through point_stat throws every ob.  The ob varname is where most of the rejects  seem to be occurring.

**Reply by JohnHalleyGotway** [reply]
I had not, but did just try it to verify 500mb temp using those 2 different input files. I do NOT see the behavior you describe. We get about the same number of pairs for 500 mb temp vs ADPUPA observations for both days:
```
   field = [
      {
        name       = "TMP";
        level      = [ "P500" ];
        cat_thresh = [ >273.0 ];
      }
   ];
```

And here's some log output for 20230203:
```
DEBUG 2: Processing TMP/P500 versus TMP/P500, for observation type ADPUPA, over region FULL, for interpolation method NEAREST(1), using 545 matched pairs.
DEBUG 3: Number of matched pairs   = 545
DEBUG 3: Observations processed    = 854123
DEBUG 3: Rejected: station id      = 0
DEBUG 3: Rejected: obs var name    = 833746
DEBUG 3: Rejected: valid time      = 0
DEBUG 3: Rejected: bad obs value   = 0
DEBUG 3: Rejected: off the grid    = 0
DEBUG 3: Rejected: topography      = 0
DEBUG 3: Rejected: level mismatch  = 19832
```
And the same output for 20230204:
```
DEBUG 2: Processing TMP/P500 versus TMP/P500, for observation type ADPUPA, over region FULL, for interpolation method NEAREST(1), using 548 matched pairs.
DEBUG 3: Number of matched pairs   = 548
DEBUG 3: Observations processed    = 910924
DEBUG 3: Rejected: station id      = 0
DEBUG 3: Rejected: obs var name    = 852197
DEBUG 3: Rejected: valid time      = 0
DEBUG 3: Rejected: bad obs value   = 0
DEBUG 3: Rejected: off the grid    = 0
DEBUG 3: Rejected: topography      = 0
DEBUG 3: Rejected: level mismatch  = 58179
```
Is there a different variable, level, or message type I should be testing other than 500mb temp?

**Comment by craigr3** [comment, marked as answer]
We were testing 2m temperature.

**Reply by JohnHalleyGotway** [reply]
Ah yes, that really helps narrow it down. I do see a difference in 2-m temp output.

I see 39188 pairs on 2023020400:
```
DEBUG 2: Processing TMP/Z2 versus TMP/Z2, for observation type ADPSFC, over region FULL, for interpolation method NEAREST(1), using 39188 matched pairs.
```
Versus 0 pairs on 2023020300:
```
DEBUG 2: Processing TMP/Z2 versus TMP/Z2, for observation type ADPSFC, over region FULL, for interpolation method NEAREST(1), using 0 matched pairs.
```
And I can see this in the ascii dump of the NetCDF files you sent:
```
grep TMP 2023020400.txt | grep ADPSFC | wc -l
39188
```
And none on the 3rd:
```
grep TMP prepbufr.2023020300.txt | grep ADPSFC | wc -l
0
```
My *GUESS* is that this has to due with the quality marker value in the PB2NC configuration file. Some PrepBUFR files set the quality marker for all *surface* observations to a value of 9. That's how they prevent those obs from being included in the data assimilation process. But it also causes this headache in the verification step.

I'd recommend that you try:
1. Reconfigure and rerun PB2NC using `quality_mark_thresh = 9;` instead of the default value of 2.
2. Reconfigure and rerun Point-Stat using `obs_quality_inc  = [ "0", "1", "2", "9" ];` to use obs that have any of these quality mark values.

This detail is mentioned at the bottom of this page:
https://dtcenter.org/community-code/model-evaluation-tools-met/input-data
```
*** NOTE: The quality flag values for surface observations in GDAS PrepBufr files are set to a value of 9 to prevent them from being used in the data assimilation process. To retain them in the PB2NC tool, the quality mark threshold in the configuration file must be set to at least 9.
```
But we should probably re-state this in the MET User's Guide as well.

Please let me know how it works out.
